timemachine/exchange/exchange_mover.py

```diff
@@ -87,108 +95,385 @@ def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff):
-        # If our system is in a clash free state, a deletion move has delta_Us typically
-        # on the order of 35kTs. However, if we start in a clashy state, we can no longer
-        # guarantee this. So it's safer to disable this micro-optimization for now.
-        delta_U_delete = -self.U_fn(coords, box, a_idxs, b_idxs)
-        delta_U_total = delta_U_delete + delta_U_insert
+# numpy version of delta_r
+def delta_r_np(ri, rj, box):
+    diff = ri - rj  # this can be either N,N,3 or B,3
+    if box is not None:
+        box_diag = np.diag(box)
+        diff -= box_diag * np.floor(diff / box_diag + 0.5)
+    return diff
+
+
+def v1_insertion(radius, center, box):
+    # (ytz): sample a point inside the sphere uniformly
+    # source: https://karthikkaranth.me/blog/generating-random-points-in-a-sphere/
+    xyz = np.random.randn(3)
+    n = np.linalg.norm(xyz)
+    xyz = xyz / n
+    # (ytz): this might not be inside the box, but it doesnt' matter
+    # since all of our distances are computed using PBCs
+    c = np.cbrt(np.random.rand())
+    new_xyz = xyz * c * radius + center
+    assert np.linalg.norm(delta_r_np(new_xyz, center, box)) < radius
+
+    return new_xyz
+
+
+def v2_insertion(radius, center, box):
+    # generate random proposals outside of the sphere but inside the box
+    while True:
+        xyz = np.random.rand(3) * np.diag(box)
+        if np.linalg.norm(delta_r_np(xyz, center, box)) >= radius:
+            return xyz
+
+
+class TIBDExchangeMove(BDExchangeMove):
+    r"""
+    Targetted Insertion and Biased Deletion Exchange Move
+
+    Insertions are targetted over two regions V1 and V2 such that V1 is a sphere whose origin
+    is at the centroid of a set of indices, and V2 = box - V1.
+
+    Deletions are biased such that each water x_i has a weight equal to exp(u_ixn_nrg(x_i, x \ x_i)).
+    """
+
+    def __init__(
+        self,
+        nb_beta: float,
+        nb_cutoff: float,
+        nb_params: NDArray,
+        water_idxs: NDArray,
+        beta: float,
+        ligand_idxs: List[int],
+        radius: float,
+    ):
+        """
+        Parameters
+        ----
+        nb_beta: float
+            nonbonded beta parameters used in direct space pme
+
+        nb_cutoff: float
+            cutoff in the nonbonded kernel
+
+        nb_params: NDArray (N,4)
+            N is the total number of atoms in the system.
+            Each 4-tuple of nonbonded parameters corresponds to (charge, sigma, epsilon, w coords)
+
+        water_idxs: NDArray (W,3)
+            W is the total number of water molecules in the system.
+            Each element is a 3-tuple denoting the index for oxygen, hydrogen, hydrogen (or whatever is
+            consistent with the nb_params)
+
+        ligand_idxs: List[int]
+            Indices corresponding to the atoms that should be used to compute the centroid
+
+        beta: float
+            Thermodynamic beta, 1/kT (in kJ/mol)
```

Comment:
nit: `(kJ/mol)^-1`

(or just accept temperature, in units kelvin, and compute beta internally... also avoids name collision with nb_beta)

Reply:
willfix